File: combine_drive_files.py
"""
Script to combine two Google Drive files (PDF, PNG, or JPEG) into one PDF.

Usage:
    python combine_drive_files.py
"""

import requests
import os
import tempfile
from pathlib import Path
from PIL import Image
import PyPDF2
from io import BytesIO
import re
import shutil

# Try to import gdown for better Google Drive support
try:
    import gdown
    GDOWN_AVAILABLE = True
except ImportError:
    GDOWN_AVAILABLE = False


def convert_google_drive_link(shareable_link):
    """
    Convert a Google Drive shareable link to a direct download link.
    
    Args:
        shareable_link: Google Drive shareable link (e.g., https://drive.google.com/file/d/FILE_ID/view?usp=sharing)
    
    Returns:
        Tuple of (file_id, direct download link)
    """
    # Extract file ID from the shareable link
    if '/file/d/' in shareable_link:
        file_id = shareable_link.split('/file/d/')[1].split('/')[0]
    elif 'id=' in shareable_link:
        file_id = shareable_link.split('id=')[1].split('&')[0]
    else:
        raise ValueError("Invalid Google Drive link format")
    
    # Return file ID and direct download link
    return file_id, f"https://drive.google.com/uc?export=download&id={file_id}"


def download_file(url, output_path):
    """
    Download a file from a URL to a local path.
    
    Args:
        url: URL to download from
        output_path: Local path to save the file
    
    Returns:
        Path to the downloaded file
    """
    print(f"Downloading file from: {url}")
    response = requests.get(url, stream=True)
    response.raise_for_status()
    
    with open(output_path, 'wb') as f:
        for chunk in response.iter_content(chunk_size=8192):
            f.write(chunk)
    
    print(f"File downloaded to: {output_path}")
    return output_path


def image_to_pdf(image_path, output_pdf_path):
    """
    Convert an image (PNG, JPEG) to PDF.
    
    Args:
        image_path: Path to the image file
        output_pdf_path: Path to save the PDF
    """
    print(f"Converting image {image_path} to PDF...")
    image = Image.open(image_path)
    
    # Convert RGBA to RGB if necessary
    if image.mode == 'RGBA':
        rgb_image = Image.new('RGB', image.size, (255, 255, 255))
        rgb_image.paste(image, mask=image.split()[3])  # Use alpha channel as mask
        image = rgb_image
    elif image.mode != 'RGB':
        image = image.convert('RGB')
    
    image.save(output_pdf_path, 'PDF', resolution=100.0)
    print(f"Image converted to PDF: {output_pdf_path}")


def combine_pdfs(pdf_paths, output_path):
    """
    Combine multiple PDF files into one.
    Decrypts with empty password when needed (e.g. bank statement PDFs).
    """
    print(f"Combining {len(pdf_paths)} PDF files...")
    pdf_merger = PyPDF2.PdfMerger()
    for pdf_path in pdf_paths:
        print(f"  Adding: {pdf_path}")
        reader = PyPDF2.PdfReader(pdf_path)
        if getattr(reader, 'is_encrypted', False):
            reader.decrypt("")
        pdf_merger.append(reader)
    pdf_merger.write(output_path)
    pdf_merger.close()
    print(f"Combined PDF saved to: {output_path}")


def download_from_google_drive(file_id, temp_dir):
    """
    Download a file from Google Drive using multiple methods.
    
    Args:
        file_id: Google Drive file ID
        temp_dir: Temporary directory to save files
    
    Returns:
        Path to the downloaded file
    """
    # First, try using gdown if available (better for Google Drive)
    if GDOWN_AVAILABLE:
        try:
            print("Attempting download using gdown library...")
            url = f"https://drive.google.com/uc?id={file_id}"
            temp_file_path = os.path.join(temp_dir, f"gdown_{file_id}")
            gdown.download(url, temp_file_path, quiet=False, fuzzy=True)
            
            # Check if file was downloaded successfully
            if os.path.exists(temp_file_path) and os.path.getsize(temp_file_path) > 0:
                print("Successfully downloaded using gdown!")
                return temp_file_path
            else:
                print("gdown download failed, trying alternative method...")
        except Exception as e:
            print(f"gdown failed: {e}, trying alternative method...")
    
    # Fallback to requests method
    session = requests.Session()
    
    # Method 1: Try direct download
    download_url = f"https://drive.google.com/uc?export=download&id={file_id}"
    response = session.get(download_url, stream=True, allow_redirects=True)
    
    # Check if we got HTML (means we need to handle virus scan warning or large file)
    content_type = response.headers.get('Content-Type', '').lower()
    content = b''
    
    # Read first chunk to check if it's HTML
    for chunk in response.iter_content(chunk_size=8192):
        content += chunk
        if len(content) > 1000:  # Read enough to detect HTML
            break
    
    # Check if response is HTML
    if b'<!doctype' in content.lower() or b'<html' in content.lower() or 'text/html' in content_type:
        print("Detected HTML response, trying alternative download method...")
        
        # Method 2: Try with confirm parameter
        download_url = f"https://drive.google.com/uc?export=download&confirm=t&id={file_id}"
        response = session.get(download_url, stream=True, allow_redirects=True)
        content = b''
        for chunk in response.iter_content(chunk_size=8192):
            content += chunk
            if len(content) > 1000:
                break
        
        # If still HTML, try to extract download link from the page
        if b'<!doctype' in content.lower() or b'<html' in content.lower():
            print("Still HTML, attempting to parse download link...")
            # Try to find the download link in the HTML
            html_content = content.decode('utf-8', errors='ignore')
            
            # Check if it's an authentication page
            if 'sign in' in html_content.lower() or 'signin' in html_content.lower():
                raise ValueError(
                    "This file requires authentication. Please make the file publicly accessible:\n"
                    "1. Right-click the file in Google Drive\n"
                    "2. Click 'Share'\n"
                    "3. Change access to 'Anyone with the link' can view\n"
                    "4. Copy the new shareable link and try again"
                )
            
            # Look for the download link pattern
            # Google Drive often has: href="/uc?export=download&id=..." or similar
            match = re.search(r'href="(/uc\?export=download[^"]+)"', html_content)
            if match:
                download_url = "https://drive.google.com" + match.group(1)
                response = session.get(download_url, stream=True, allow_redirects=True)
            else:
                # Try alternative pattern
                match = re.search(r'id="uc-download-link"[^>]*href="([^"]+)"', html_content)
                if match:
                    download_url = "https://drive.google.com" + match.group(1)
                    response = session.get(download_url, stream=True, allow_redirects=True)
                else:
                    # Last resort: try the alternative API endpoint
                    download_url = f"https://drive.google.com/uc?id={file_id}&export=download"
                    response = session.get(download_url, stream=True, allow_redirects=True)
    
    # Determine file extension from Content-Type or default
    content_type = response.headers.get('Content-Type', '').lower()
    file_extension = '.tmp'
    
    if 'pdf' in content_type:
        file_extension = '.pdf'
    elif 'png' in content_type:
        file_extension = '.png'
    elif 'jpeg' in content_type or 'jpg' in content_type:
        file_extension = '.jpg'
    
    # Create temporary file
    temp_file = tempfile.NamedTemporaryFile(delete=False, dir=temp_dir, suffix=file_extension)
    temp_file_path = temp_file.name
    temp_file.close()
    
    # Download the entire file
    with open(temp_file_path, 'wb') as f:
        # Write the content we already read
        if content:
            f.write(content)
        # Write the rest
        for chunk in response.iter_content(chunk_size=8192):
            f.write(chunk)
    
    return temp_file_path


def process_file(google_drive_link, temp_dir):
    """
    Download a file from Google Drive and convert it to PDF if needed.
    
    Args:
        google_drive_link: Google Drive shareable link
        temp_dir: Temporary directory to save files
    
    Returns:
        Path to the PDF file (original or converted)
    """
    # Convert Google Drive link to get file ID
    file_id, _ = convert_google_drive_link(google_drive_link)
    
    # Download the file
    temp_file_path = download_from_google_drive(file_id, temp_dir)
    
    # Check actual file type by reading first bytes
    with open(temp_file_path, 'rb') as f:
        file_header = f.read(4)
    
    # Determine file type from magic bytes
    if file_header.startswith(b'%PDF'):
        # It's a PDF
        print(f"File is a PDF: {temp_file_path}")
        return temp_file_path
    elif file_header.startswith(b'\x89PNG'):
        # It's a PNG
        print(f"File is a PNG: {temp_file_path}")
        pdf_path = os.path.splitext(temp_file_path)[0] + '.pdf'
        image_to_pdf(temp_file_path, pdf_path)
        os.remove(temp_file_path)  # Remove original image
        return pdf_path
    elif file_header.startswith(b'\xff\xd8\xff'):
        # It's a JPEG
        print(f"File is a JPEG: {temp_file_path}")
        pdf_path = os.path.splitext(temp_file_path)[0] + '.pdf'
        image_to_pdf(temp_file_path, pdf_path)
        os.remove(temp_file_path)  # Remove original image
        return pdf_path
    else:
        # Try reading more bytes to see if it's HTML
        with open(temp_file_path, 'rb') as f:
            more_content = f.read(100)
        if b'<!doctype' in more_content.lower() or b'<html' in more_content.lower():
            raise ValueError("Google Drive returned an HTML page instead of the file. The file may be too large or require permission. Please make sure the file is publicly accessible or try a different link format.")
        raise ValueError(f"Unsupported file type. File header: {file_header}")


def main():
    """
    Main function to combine two Google Drive files into one PDF.
    """
    print("=" * 60)
    print("Google Drive File Combiner")
    print("=" * 60)
    print()
    
    # Get the project directory (where this script is located)
    project_dir = Path(__file__).parent.absolute()
    
    # Create temp and outputs directories within the project
    temp_dir = project_dir / "temp"
    outputs_dir = project_dir / "outputs"
    
    # Create directories if they don't exist
    temp_dir.mkdir(exist_ok=True)
    outputs_dir.mkdir(exist_ok=True)
    
    print(f"Using project temp directory: {temp_dir}")
    print(f"Outputs will be saved to: {outputs_dir}")
    print()
    
    # Get Google Drive links from user
    print("Enter the first Google Drive link (PDF, PNG, or JPEG):")
    link1 = input().strip()
    
    print("\nEnter the second Google Drive link (PDF, PNG, or JPEG):")
    link2 = input().strip()
    
    # Get output filename
    print("\nEnter output PDF filename (default: combined_output.pdf):")
    output_filename = input().strip()
    if not output_filename:
        output_filename = "combined_output.pdf"
    if not output_filename.endswith('.pdf'):
        output_filename += '.pdf'
    
    # Full path for output file
    output_path = outputs_dir / output_filename
    
    try:
        # Process both files
        pdf_paths = []
        
        print("\n" + "=" * 60)
        print("Processing first file...")
        print("=" * 60)
        pdf1 = process_file(link1, str(temp_dir))
        pdf_paths.append(pdf1)
        
        print("\n" + "=" * 60)
        print("Processing second file...")
        print("=" * 60)
        pdf2 = process_file(link2, str(temp_dir))
        pdf_paths.append(pdf2)
        
        # Combine PDFs
        print("\n" + "=" * 60)
        print("Combining files...")
        print("=" * 60)
        combine_pdfs(pdf_paths, str(output_path))
        
        print("\n" + "=" * 60)
        print(f"SUCCESS! Combined PDF saved as: {output_path}")
        print("=" * 60)
        
    except Exception as e:
        print(f"\nERROR: {str(e)}")
        import traceback
        traceback.print_exc()
    
    finally:
        # Clean up temporary files (but keep the temp directory)
        print(f"\nCleaning up temporary files...")
        for file in temp_dir.glob("*"):
            try:
                if file.is_file():
                    file.unlink()
            except Exception as e:
                print(f"Warning: Could not delete {file}: {e}")
        print("Done!")


if __name__ == "__main__":
    main()

File: combine_drive_files.js
/**
 * Combine two Google Drive files (PDF, PNG, or JPEG) into one PDF.
 * Same purpose as combine_drive_files.py, in Node.js.
 *
 * Usage:
 *   node combine_drive_files.js
 *   (prompts for link1, link2, output filename)
 *
 *   node combine_drive_files.js <link1> <link2> [output.pdf]
 *
 * Requires Node 18+ (for fetch). Encrypted PDFs will fail (pdf-lib cannot decrypt).
 */

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import { createInterface } from 'readline';
import { PDFDocument } from 'pdf-lib';

const __dirname = path.dirname(fileURLToPath(import.meta.url));

// --- Link parsing ---

function extractFileId(shareableLink) {
  const link = (shareableLink || '').trim();
  if (link.includes('/file/d/')) {
    return link.split('/file/d/')[1].split('/')[0];
  }
  if (link.includes('id=')) {
    return link.split('id=')[1].split('&')[0];
  }
  throw new Error('Invalid Google Drive link format');
}

// --- Download from Google Drive (same strategies as Python script) ---

const DRIVE_URLS = (fileId) => [
  `https://drive.google.com/uc?export=download&id=${fileId}`,
  `https://drive.google.com/uc?export=download&confirm=t&id=${fileId}`,
  `https://drive.google.com/uc?id=${fileId}&export=download`,
];

function isHtml(buffer) {
  if (!buffer || buffer.length < 8) return false;
  const start = buffer.slice(0, 256).toString('utf8').toLowerCase();
  return start.includes('<!doctype') || start.includes('<html');
}

function isSignInPage(html) {
  return /sign\s*in|signin/i.test(html);
}

function parseDownloadLinkFromHtml(html) {
  const m = html.match(/href="(\/uc\?export=download[^"]+)"/);
  if (m) return 'https://drive.google.com' + m[1];
  const m2 = html.match(/id="uc-download-link"[^>]*href="([^"]+)"/);
  if (m2) return 'https://drive.google.com' + m2[1];
  return null;
}

async function downloadFromGoogleDrive(fileId) {
  let lastError = null;
  for (const url of DRIVE_URLS(fileId)) {
    try {
      const res = await fetch(url, { redirect: 'follow' });
      if (!res.ok) continue;
      const contentType = (res.headers.get('Content-Type') || '').toLowerCase();
      if (contentType.includes('text/html')) {
        const html = await res.text();
        if (isSignInPage(html)) {
          throw new Error(
            'File requires authentication. Make the file publicly accessible: Share → Anyone with the link can view.'
          );
        }
        const downloadUrl = parseDownloadLinkFromHtml(html);
        if (downloadUrl) {
          const res2 = await fetch(downloadUrl, { redirect: 'follow' });
          if (res2.ok) {
            const buf = Buffer.from(await res2.arrayBuffer());
            if (!isHtml(buf)) return buf;
          }
        }
        continue;
      }
      const buf = Buffer.from(await res.arrayBuffer());
      if (buf.length === 0) continue;
      if (isHtml(buf)) continue;
      return buf;
    } catch (e) {
      lastError = e;
    }
  }
  throw lastError || new Error(`Download failed for ${fileId}. Make the file publicly accessible (Anyone with the link).`);
}

// --- Magic byte detection (same as Python) ---

function sniffFileKind(buffer) {
  if (!buffer || buffer.length < 4) return 'unknown';
  // PDF: %PDF-
  if (buffer[0] === 0x25 && buffer[1] === 0x50 && buffer[2] === 0x44 && buffer[3] === 0x46 && buffer[4] === 0x2d) return 'pdf';
  // JPEG: FF D8 FF
  if (buffer[0] === 0xff && buffer[1] === 0xd8 && buffer[2] === 0xff) return 'jpg';
  // PNG: 89 50 4E 47 0D 0A 1A 0A
  if (buffer.length >= 8 &&
      buffer[0] === 0x89 && buffer[1] === 0x50 && buffer[2] === 0x4e && buffer[3] === 0x47 &&
      buffer[4] === 0x0d && buffer[5] === 0x0a && buffer[6] === 0x1a && buffer[7] === 0x0a) return 'png';
  return 'unknown';
}

// --- Image to single-page PDF (pdf-lib) ---

async function imageToPdfBytes(imageBuffer, kind) {
  const doc = await PDFDocument.create();
  const page = doc.addPage();
  const { width, height } = page.getSize();
  if (kind === 'png') {
    const img = await doc.embedPng(imageBuffer);
    const scale = Math.min(width / img.width, height / img.height);
    page.drawImage(img, { x: 0, y: 0, width: img.width * scale, height: img.height * scale });
  } else {
    const img = await doc.embedJpg(imageBuffer);
    const scale = Math.min(width / img.width, height / img.height);
    page.drawImage(img, { x: 0, y: 0, width: img.width * scale, height: img.height * scale });
  }
  return await doc.save();
}

// --- Combine PDFs (merge in order; encrypted PDFs will throw) ---

async function combinePdfs(pdfBuffers) {
  const mergedDoc = await PDFDocument.create();
  for (let i = 0; i < pdfBuffers.length; i++) {
    const buf = pdfBuffers[i];
    const srcDoc = await PDFDocument.load(buf);
    const pageCount = srcDoc.getPageCount();
    if (pageCount === 0) throw new Error(`File ${i + 1} has no pages.`);
    const copied = await mergedDoc.copyPages(srcDoc, [...Array(pageCount)].map((_, j) => j));
    copied.forEach((p) => mergedDoc.addPage(p));
  }
  return await mergedDoc.save();
}

// --- Process one Drive link: download → detect type → convert image to PDF if needed → return PDF bytes ---

async function processFile(link, _tempDir) {
  const fileId = extractFileId(link);
  const buffer = await downloadFromGoogleDrive(fileId);
  const kind = sniffFileKind(buffer);
  if (kind === 'pdf') {
    console.log('  File is a PDF');
    return buffer;
  }
  if (kind === 'png') {
    console.log('  File is a PNG, converting to PDF...');
    return await imageToPdfBytes(buffer, 'png');
  }
  if (kind === 'jpg') {
    console.log('  File is a JPEG, converting to PDF...');
    return await imageToPdfBytes(buffer, 'jpg');
  }
  if (buffer.length >= 100 && isHtml(buffer)) {
    throw new Error('Google Drive returned an HTML page. File may be too large or require permission. Make the file publicly accessible.');
  }
  throw new Error(`Unsupported file type. First bytes: ${buffer.slice(0, 8).toString('hex')}`);
}

/** Combine two Drive links (PDF/PNG/JPEG) into one PDF. Returns merged PDF as Uint8Array. */
export async function combineTwoLinks(link1, link2) {
  const pdf1 = await processFile(link1, '');
  const pdf2 = await processFile(link2, '');
  return await combinePdfs([pdf1, pdf2]);
}

// --- Main ---

function ensureDir(dir) {
  if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });
}

function question(readline, prompt) {
  return new Promise((resolve) => readline.question(prompt, resolve));
}

async function mainInteractive() {
  const rl = createInterface({ input: process.stdin, output: process.stdout });
  const link1 = (await question(rl, 'Enter the first Google Drive link (PDF, PNG, or JPEG): ')).trim();
  const link2 = (await question(rl, '\nEnter the second Google Drive link (PDF, PNG, or JPEG): ')).trim();
  let outputName = (await question(rl, '\nEnter output PDF filename (default: combined_output.pdf): ')).trim();
  rl.close();
  if (!outputName) outputName = 'combined_output.pdf';
  if (!outputName.endsWith('.pdf')) outputName += '.pdf';
  return { link1, link2, outputName };
}

async function main() {
  const projectDir = __dirname;
  const tempDir = path.join(projectDir, 'temp');
  const outputsDir = path.join(projectDir, 'outputs');
  ensureDir(tempDir);
  ensureDir(outputsDir);

  let link1, link2, outputName;
  if (process.argv.length >= 4) {
    link1 = process.argv[2];
    link2 = process.argv[3];
    outputName = process.argv[4] || 'combined_output.pdf';
    if (!outputName.endsWith('.pdf')) outputName += '.pdf';
  } else {
    console.log('Google Drive File Combiner (Node.js)\n');
    const answers = await mainInteractive();
    link1 = answers.link1;
    link2 = answers.link2;
    outputName = answers.outputName;
  }

  const outputPath = path.join(outputsDir, outputName);

  console.log('\nProcessing first file...');
  const pdf1 = await processFile(link1, tempDir);
  console.log('\nProcessing second file...');
  const pdf2 = await processFile(link2, tempDir);
  console.log('\nCombining PDFs...');
  const merged = await combinePdfs([pdf1, pdf2]);
  fs.writeFileSync(outputPath, merged);
  console.log(`\nSUCCESS! Combined PDF saved as: ${outputPath}`);
}

const isRunDirectly =
  process.argv[1] && path.resolve(process.argv[1]) === path.resolve(__dirname, 'combine_drive_files.js');
if (isRunDirectly) {
  main().catch((e) => {
    console.error('\nERROR:', e.message);
    process.exit(1);
  });
}
